dag_test.go

```diff
@@ -94,6 +97,89 @@ func TestUndirectedTopologicalSort(t *testing.T) {
 	}
 }
 
+func TestDirectedStableTopologicalSort(t *testing.T) {
+	tests := map[string]struct {
+		vertices      []int
+		edges         []Edge[int]
+		expectedOrder []int
+		shouldFail    bool
+	}{
+		"graph with 5 vertices": {
+			vertices: []int{1, 2, 3, 4, 5},
+			edges: []Edge[int]{
+				{Source: 1, Target: 2},
+				{Source: 1, Target: 3},
+				{Source: 2, Target: 3},
+				{Source: 2, Target: 4},
+				{Source: 2, Target: 5},
+				{Source: 3, Target: 4},
+				{Source: 4, Target: 5},
+			},
+			expectedOrder: []int{1, 2, 3, 4, 5},
+		},
+		"graph with many possible topological orders": {
+			vertices: []int{1, 2, 3, 4, 5, 6, 10, 20, 30, 40, 50, 60},
+			edges: []Edge[int]{
+				{Source: 1, Target: 10},
+				{Source: 2, Target: 20},
+				{Source: 3, Target: 30},
+				{Source: 4, Target: 40},
+				{Source: 5, Target: 50},
+				{Source: 6, Target: 60},
+			},
+			expectedOrder: []int{1, 2, 3, 4, 5, 6, 10, 20, 30, 40, 50, 60},
+		},
```

Comment:
This is the interesting test case. It fails using `TopologicalSort` and passes using `StableTopologicalSort`.

Reply:
Essentially, this becomes breadth-first: all of the level 0s (1,2,3,4,5,6), then all of the level 1s (10,20,30,40,50,60). It uses the `less` to decide "1 before 2", "2 before 3", etc., and then, because 1 always has to be before 10, you end up with 10,20,etc. on the same level, and uses the `less` to decide "10 before 20", "20 before 30", etc.

Essentially, `less` decides among candidates on the _same_ level, but is silent on choosing between levels. Both of the following, though, would be equally valid sorted outputs:

* `1,2,3,4,5,6,10,20,30,40,50,60`
* `1,10,2,20,3,30,4,40,5,50,6,60`

Is there a way to use the `less` function to get depth-first, e.g. `1,10,2,20,...`? I don't think it would be worth holding this up to make that a possibility, but it is part of sorting, and it would be a pity to have to change the API later. Or maybe that is just a separate option? 🤷‍♂️

Reply:
The topological order is determined using [Kahn's algorithm](https://www.interviewcake.com/concept/java/topological-sort), which essentially _is_ a breadth-first-style algorithm. So that's by design.

There is no way to get a DFS-style ordering at the moment. Is there a particular reason why a DFS ordering would be preferable over a BFS ordering in a topological sort? And if there is, wouldn't a `DFS` call do the job? Quite frankly, I'm not aware of all the use cases and requirements of the topological sort.

Reply:
Nope not that I'm aware of. I certainly don't have that use case now. I'm just thinking ahead as to whether having it would end up changing the API and giving you a bigger headache later. If the answer is, it doesn't matter, then it doesn't matter. 😃

Reply:
Agree with you.

It would be better to have "traversing" and "sorting" expressed in separate APIs. `StableTopologicalSort` is good enough.

If a user wants to customize their own, more complicated logic, they can extend it by defining their own traversal rules, but this doesn't seem to be a common case. (I guess)